paddle/operators/deconv2d_op.h

```diff
@@ -0,0 +1,251 @@
+      // batch with size (M, H * W)
+      Tensor input_batch = input->Slice<T>(i, i + 1).Resize(input_matrix_shape);
+      // filter size: (M, C * K_H * K_W)
+
+      // output size: (C, O_H, O_W)
+      Tensor output_batch = output->Slice<T>(i, i + 1).Resize(output_shape);
+
+      // col_matrix = filter * input_batch
+      // of shape (C * K_H * K_W, H * W)
+      math::matmul<Place, T>(context.device_context(), filter, true,
+                             input_batch, false, T(1.0), &col_matrix, T(0.0));
+      col2im(context.device_context(), output_batch, col, strides[0],
+             strides[1], 0, 0);
+    }
+  }
+};
+
+template <typename Place, typename T>
+class GemmDeconvGrad2DKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    const Tensor* input = context.Input<Tensor>("Input");
+    const Tensor* output_grad =
+        context.Input<Tensor>(framework::GradVarName("Output"));
+
+    // For filter, we do not use const pointer b/c we will do reshape
+    // but we should avoid modifying its value
+    Tensor filter = *context.Input<Tensor>("Filter");
+
+    Tensor* input_grad =
+        context.Output<Tensor>(framework::GradVarName("Input"));
+    Tensor* filter_grad =
+        context.Output<Tensor>(framework::GradVarName("Filter"));
+
+    std::vector<int> strides = context.Attr<std::vector<int>>("strides");
+    // Actually, no paddings and groups allowed in deconv
+    std::vector<int> paddings = context.Attr<std::vector<int>>("paddings");
+
+    int N = input->dims()[0];
+    int M = input->dims()[1];
+    int H = input->dims()[2];
+    int W = input->dims()[3];
+
+    int K_H = filter.dims()[2];
+    int K_W = filter.dims()[3];
+
+    int C = output_grad->dims()[1];  // output channels
+    int O_H = output_grad->dims()[2];
+    int O_W = output_grad->dims()[3];
+
+    // Only im2col functor required for bp to get to the right shape
+    paddle::operators::math::Im2ColFunctor<
+        paddle::operators::math::ColFormat::kCFO, Place, T>
+        im2col;
+
+    // use col_shape in the im2col and col2im calculation
+    DDim col_shape = {C, K_H, K_W, H, W};
+
+    // use col_matrix_shape in the gemm calculation
+    DDim col_matrix_shape_f = {C * H * W, K_H * K_W};
+
+    Tensor col;
+    col.mutable_data<T>(col_shape, context.GetPlace());
+    // col_matrix shares the same piece of data with col,
+    // but will be reshaped into a two-dimensional matrix shape
+    // to call the matrix multiplication interface.
+
+    DDim output_shape = {C, O_H, O_W};
+    DDim input_matrix_shape = {M, H * W};
+
+    DDim filter_matrix_shape = {M, C * K_H * K_W};
+    filter.Resize(filter_matrix_shape);
+
+    // deconvolution grad on input:
+    // im2col + gemm (similar to conv-forward)
+    // input need to compute gradient
+    if (input_grad) {
+      Tensor col_matrix = col;
+      DDim col_matrix_shape = {C * K_H * K_W, H * W};
```

Comment:
See above comment. I would prefer the more explicit:
```
Tensor col_matrix;
col_matrix.ShareDataWith(col);
```

Reply:
Done.